GSW/testUtility/testNS2.pyw

```diff
@@ -26,87 +26,148 @@ commandField = Entry(commandFrame, width=30, borderwidth=3, font=('Consolas',10)
 
 # log frame
 logFrame = LabelFrame(root, text='Log', padx=5, pady=5)
-log = tkScrolledText.ScrolledText(logFrame, state='disabled', width=50, height=20, font=('Consolas',10))
+log = tkScrolledText.ScrolledText(logFrame, state='disabled', width=50, height=20, font=('Consolas',10),  wrap=WORD)
 
 # monitor frame
 monitorFrame = LabelFrame(root, text='NS2 Output', padx=5, pady=5)
 monitor = tkScrolledText.ScrolledText(monitorFrame, state='disabled', width=80, height=30, font=('Consolas',10))
+autoScroll = BooleanVar(value=True)
+check_autoScroll = Checkbutton(monitorFrame, text='Auto Scroll',variable=autoScroll, onvalue=True, offvalue=False)
+check_autoScroll.select()
+
+# file frame
+fileFrame = LabelFrame(root, text='Save to File', padx=5, pady=5)
+fileNameField = Entry(fileFrame, width=30, borderwidth=3)
+fileNameField.insert(0, 'filename')
 
 # serial setup frame
-serialFrame = LabelFrame(root, text='Setup Serial', padx = 5, pady = 5)
+serialFrame = LabelFrame(root, text='Setup Serial', padx=5, pady=5)
 portField = Entry(serialFrame, width=30, borderwidth=3)
 portField.insert(0, 'COM3')
 
 
 # ==== define functions and buttons ====
-# update log
-def updateLog(text):
+# update log #
+def updateLog(text): # writes text to the log
     log.configure(state ='normal')
-    log.insert(INSERT, text + '\n')
+    log.insert(END, text + '\r\n')
     log.configure(state ='disabled')
     log.see(END)
 
-# send command
-def sendCommand(event=None):
+# send command #
+def sendCommand(event=None): # sends a command over serial
+    if not portOpen:
+        updateLog('No connection!')
+        return
     command = commandField.get()
     if command:
         teensy.write(bytes(command, 'utf-8'))
         updateLog('Command Sent: ' + command)
         commandField.delete(0, END)
-button_sendCommand = Button(commandFrame, text='Send', command=sendCommand)
-# bind the enter key to send command
-root.bind('<Return>', sendCommand)
 
-# read serial
-def readSerial():
+# read serial #
+def readSerial(): # reads incoming communication from NS2 via the open serial port
     data = str(teensy.readline())[2:-1]
     if data:
         monitor.configure(state ='normal')
-        monitor.insert(INSERT, codecs.decode(data, "unicode_escape"))
+        monitor.insert(END, codecs.decode(data, "unicode_escape")) # insert data in monitor
         monitor.configure(state ='disabled')
-        monitor.see(END)
+        if autoScroll.get():
+            monitor.see(END) 
 
 # scan for serial ports
-def scanPorts():
+def scanPorts(): # finds and lists all available com ports
     availablePorts = list_ports.comports()
     updateLog('==== Found Serial Ports: ====')
     for port in availablePorts:
         updateLog('- ' + port.description)
```

Comment:
perhaps print a message saying "no ports" if there are none?
![image](https://user-images.githubusercontent.com/49298404/112385408-33be2280-8cb5-11eb-9ea6-e89a66d80fa3.png)

Reply:
how do you even have no serial ports?? I'll add that in.

Reply:
Lmfao